### PR TITLE
Fixed #29501 -- Allowed dbshell to pass options to underlying tool.

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket](https://code.djangoproject.com/ticket/29501)

### CODE REVIEW

**Comment by adamchainz** [comment]
Requested your review @claudep because you've reviewed a few dbshell PR's in the past :)

File: django/db/backends/base/client.py [hunk 1]
```diff
@@ -8,5 +8,5 @@ def __init__(self, connection):
         # connection is an instance of BaseDatabaseWrapper.
         self.connection = connection
 
-    def runshell(self):
+    def runshell(self, parameters):
```
Comment (claudep) [inline django/db/backends/base/client.py]: Do you really want to make the `parameters` parameter mandatory? This might break third-party backends.

File: docs/ref/django-admin.txt [hunk 2]
```diff
@@ -226,6 +226,17 @@ program manually.
 
 Specifies the database onto which to open a shell. Defaults to ``default``.
 
+.. django-admin-option:: -- <args>
+
```
Comment (claudep) [inline docs/ref/django-admin.txt]: A `versionadded` directive is missing.

File: docs/ref/django-admin.txt [hunk 3]
```diff
@@ -226,6 +226,17 @@ program manually.
...
+Any arguments following a ``--`` divider will be passed on to the underlying
+program. For example if using SQLite, you can use its ``-version`` flag like
+so:
+
+.. console::
+
+    $ python manage.py dbshell -- -version
+    3.24.0 2018-06-04 14:10:15 95fbac39baaab1c3a84fdfc82ccb7f42398b2e92f18a2a57bce1d4a713cbaapl
```
Comment (claudep) [inline docs/ref/django-admin.txt]: What about providing an example with a SQL command, as the original ticket use case was targeting?

File: django/db/backends/base/client.py (hunk 1 above)
Comment (adamchainz) [reply django/db/backends/base/client.py]: If we don't make it mandatory, backends might not spot the need to update, breaking `dbshell -- --foo`? I thought we make no compat guarantees for third-party backends.

File: docs/ref/django-admin.txt (hunk 3 above)
Comment (adamchainz) [reply docs/ref/django-admin.txt]: Added

File: docs/ref/django-admin.txt (hunk 2 above)
Comment (adamchainz) [reply docs/ref/django-admin.txt]: Added

File: django/db/backends/base/client.py (hunk 1 above)
Comment (claudep) [reply django/db/backends/base/client.py]: No guarantees doesn't mean no efforts :-)  However let others chime in and tell their opinion on this.

Comment (adamchainz) [reply django/db/backends/base/client.py]: If we don't make it mandatory, when the `handle()` calls `runshell` it will crash with `TypeError`.

Comment (carltongibson) [reply django/db/backends/base/client.py]: Sounds like there's no choice then? 

At least an additional note in 3.1.txt `Database backend API` section. (I've always imagined backend maintainers keep an eye on that for exactly this reason...)

File: tests/dbshell/test_sqlite.py [hunk 4]
```diff
@@ -29,3 +32,13 @@ def test_path_name(self):
...
             )
+
+    def test_parameters(self):
+        with mock.patch.dict(
+            connection.settings_dict,
+            {'NAME': Path('test.db.sqlite3')},
+        ):
+            self.assertEqual(
+                self._run_dbshell(['-help']),
+                ['sqlite3', 'test.db.sqlite3', '-help'],
```
Comment (carltongibson) [inline tests/dbshell/test_sqlite.py]: `'--help'`

File: docs/ref/django-admin.txt [hunk 5]
```diff
@@ -226,6 +226,23 @@ program manually.
...
+
+.. console::
+
+    $ python manage.py dbshell -- -e "SELECT NOW()"
+    +---------------------+
+    | NOW()               |
+    +---------------------+
+    | 2020-02-29 16:28:49 |
+    +---------------------+
+
```
Comment (carltongibson) [inline docs/ref/django-admin.txt]: I just wonder if we can help folks a little more here, showing them the basic option for each supported DB? 

(`-c`, `--cmd`, `...😬...`) 

Danger is it gets too verbose. Would certainly need a "See your DB's ..."

Comment (adamchainz) [reply docs/ref/django-admin.txt]: I've gone halfway with a Postgres example first, given it's like 90% of the community.

File: django/db/backends/base/client.py (hunk 1 above)
Comment (adamchainz) [reply django/db/backends/base/client.py]: Added there - wasn't aware of that docs section, thanks Carlton.

File: tests/dbshell/test_sqlite.py (hunk 4 above)
Comment (carltongibson) [reply tests/dbshell/test_sqlite.py]: OK, both seem to work testing it... 👍

**Comment by adamchainz** [comment]
I think this is ready for merge?

**Comment by felixxm** [comment]
@adamchainz I will take a look in this week :eyes: , please be patient.

File: django/core/management/commands/dbshell.py [hunk 6]
```diff
@@ -15,11 +15,13 @@ def add_arguments(self, parser):
...
             help='Nominates a database onto which to open a shell. Defaults to the "default" database.',
         )
+        parameters = parser.add_argument_group('parameters', prefix_chars='--')
+        parameters.add_argument('parameters', nargs='*')
 
     def handle(self, **options):
         connection = connections[options['database']]
         try:
-            connection.client.runshell()
+            connection.client.runshell(options['parameters'])
```
Comment (felixxm) [inline django/core/management/commands/dbshell.py]: I wonder if we should suppress `subprocess.CalledProcessError` when arguments are not correct, to get e.g.
```
/usr/lib/postgresql/10/bin/psql: nieznana opcja '--commandasdasd'
Try "psql --help" for more information.
```
instead of
```
/usr/lib/postgresql/10/bin/psql: nieznana opcja '--commandasdasd'
Try "psql --help" for more information.
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    main()
  File "manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "django/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "django/django/core/management/base.py", line 328, in run_from_argv
    self.execute(*args, **cmd_options)
  File "django/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "django/django/core/management/commands/dbshell.py", line 25, in handle
    connection.client.runshell(options['parameters'])
  File "django/django/db/backends/postgresql/client.py", line 55, in runshell
    self.runshell_db(self.connection.get_connection_params(), parameters)
  File "django/django/db/backends/postgresql/client.py", line 49, in runshell_db
    subprocess.run(args, check=True, env=subprocess_env)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['psql', '-U', 'djangoticket', '-h', 'localhost', '-p', '5432', 'djangoticket', '--commandasdasd']' returned non-zero exit status 1.
```

We can also re-raise it as a `CommandError`:
```
/usr/lib/postgresql/10/bin/psql: nieznana opcja '--commandasdasd'
Try "psql --help" for more information.
CommandError: Command '['psql', '-U', 'djangoticket', '-h', 'localhost', '-p', '5432', 'djangoticket', '--commandasdasd']' returned non-zero exit status 1.
```

**Comment by felixxm** [comment]
@adamchainz Thanks for this patch :+1: I pushed edits to docs. What do you think about [suppressing](https://github.com/django/django/pull/12508#discussion_r393613017) `subprocess`' exceptions?

**Comment by pope1ni** [comment]
I think suppressing the exception makes sense. Django is passing the arguments through and the error is raised by the external process which has already thrown messages to the console.

I was going to say reraising was not needed, but if scripted it could be handy to know which exact command failed.

File: django/core/management/commands/dbshell.py (hunk 6 above)
Comment (felixxm) [reply django/core/management/commands/dbshell.py]: I'm in favor of re-raising `subprocess.CalledProcessError` as a `CommandError`, e.g.
```python
    def handle(self, **options):
        connection = connections[options['database']]
        try:
            connection.client.runshell(options['parameters'])
        except OSError:
            # Note that we're assuming OSError means that the client program
            # isn't installed. There's a possibility OSError would be raised
            # for some other reason, in which case this error message would be
            # inaccurate. Still, this message catches the common case.
            raise CommandError(
                'You appear not to have the %r program installed or on your path.' %
                connection.client.executable_name
            )
        except subprocess.CalledProcessError as e:
            raise CommandError(
                '"%s" returned non-zero exit status %s.' % (
                    ' '.join(e.cmd),
                    e.returncode),
            )
```
which will end with:
```
/usr/lib/postgresql/10/bin/psql: nieznana opcja '--commandasdasd'
Try "psql --help" for more information.
CommandError: "psql -U djangoticket -h localhost -p 5432 djangoticket --asdasdad" returned non-zero exit status 1.
```

**Comment by adamchainz** [comment]
Sorry about my impatience @felixxm 🙈

Agree with both that suppressing the exception makes sense. It already breaks some workflows before passing options through.

The reraise of `CommandError` would work, but I'm not sure it's necessary. I've instead changed the PR right now to use this pattern:

```python
proc = subprocess.run(...)
sys.exit(proc.returncode)
```

This means Django is "transparent" around the command - it won't create any output but still exits with the same code. It's much more like its behaviour using `os.exec*` which was removed in bf5382c6e549d78da9f7029cd86686459b52eaed due to windows concerns. And then `dbshell` can (again) be a complete drop-in replacement in scripts.

**Comment by felixxm** [comment]
> This means Django is "transparent" around the command - it won't create any output but still exits with the same code. It's much more like its behaviour using `os.exec*` which was removed in [bf5382c](https://github.com/django/django/commit/bf5382c6e549d78da9f7029cd86686459b52eaed) due to windows concerns. And then `dbshell` can (again) be a complete drop-in replacement in scripts.

This approach breaks usage with `call_command()`, I think we need to stay with suppression `subprocess.CalledProcessError`'s (or re-raise them as `CommandError`'s).

**Comment by adamchainz** [comment]
> This approach breaks usage with `call_command()`, I think we need to stay with suppression `subprocess.CalledProcessError`'s (or re-raise them as `CommandError`'s).

I am not sure use of `dbshell` with `call_command` is common. It would not have worked Django< 1.6 at least due to use of `os.exec*` previously, since that replaces the process and stops Python executing entirely. And anyway like this, one can catch `SystemExit` instead of `CommandError`.

I think it's important to preserve the return code, especially when adding the ability to use custom flags. Which return code was used can carry extra meaning. I've made the `sys.exit` conditional on the return code not being 0 so at least testing with `except SystemExit` catches abnormal conditions only.

**Comment by adamchainz** [comment]
I've also changed to `FileNotFoundError` in `dbshell.py`, since Python 3 differentiates between different kinds of `OSError`, and that's the one relating to missing commands.

**Comment by carltongibson** [comment]
Management commands are expected to raise `CommandError`:

> raising `CommandError` exception (with a sensible description of the error) is the preferred way to indicate that something has gone wrong in the execution of a command.

I think we need to stick with that.

**Comment by adamchainz** [comment]
Okay done. To preserve the return code I added the `returncode` argument to `CommandError`, which also seems like it could be useful for other scripting sitautions.

File: django/core/management/base.py [hunk 7]
```diff
@@ -326,16 +328,16 @@ def run_from_argv(self, argv):
         handle_default_options(options)
         try:
             self.execute(*args, **cmd_options)
-        except Exception as e:
-            if options.traceback or not isinstance(e, CommandError):
```
Comment (adamchainz) [inline django/core/management/base.py]: I'm not crazy - this is `isinstance` was redundant right?

File: django/core/management/commands/dbshell.py [hunk 8]
```diff
@@ -15,17 +17,27 @@ def add_arguments(self, parser):
...
 
     def handle(self, **options):
         connection = connections[options['database']]
         try:
-            connection.client.runshell()
-        except OSError:
-            # Note that we're assuming OSError means that the client program
-            # isn't installed. There's a possibility OSError would be raised
-            # for some other reason, in which case this error message would be
-            # inaccurate. Still, this message catches the common case.
```
Comment (felixxm) [inline django/core/management/commands/dbshell.py:27]: This is an unrelated cleanup. Please move to the separate commit.

File: django/core/management/base.py (hunk 7 above)
Comment (felixxm) [reply django/core/management/base.py:330]: This is also an unrelated cleanup that should be moved to a separate commit.

File: django/core/management/base.py [hunk 9]
```diff
@@ -326,16 +328,16 @@ def run_from_argv(self, argv):
...
+        except CommandError as e:
+            if options.traceback:
                 raise
 
             # SystemCheckError takes care of its own formatting.
             if isinstance(e, SystemCheckError):
                 self.stderr.write(str(e), lambda x: x)
             else:
                 self.stderr.write('%s: %s' % (e.__class__.__name__, e))
-            sys.exit(1)
```
Comment (felixxm) [inline django/core/management/base.py:338]: We shouldn't mix this cleanup with a new feature. Please move it to a separate commit also test is missing.

**Comment by adamchainz** [comment]
Apologies for my unsplit commits. I think I've done it with good enough naming now.

**Comment by felixxm** [comment]
@adamchainz Thanks :+1: I added tests for missing `dbshell` executable (in a separate commit), added test for `CommandError.returncode`, and pushed minor edits.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by adamchainz** [comment]
All LGTM

File: docs/releases/3.1.txt [hunk 10]
```diff
@@ -501,6 +508,9 @@ backends.
   yields a cursor and automatically closes the cursor and connection upon
   exiting the ``with`` statement.
 
+* ``DatabaseClient.runshell()`` now requires an additional ``parameters``
+  argument for extra arguments to pass on to the command-line client.
```
Comment (carltongibson) [inline docs/releases/3.1.txt:512]: Just the one comment/question: `parameters` is a list right? Is it worth saying it's a list here?

Comment (adamchainz) [reply docs/releases/3.1.txt:512]: Rephrased
